### PR TITLE
IDGEN-79 Link Auto Generation Options EndPoints to OWA

### PR DESCRIPTION
**Pull request by andela-dbamidele**
## JIRA TICKET NAME:

[IDGEN-79 Link Auto Generation Options EndPoints to OWA](https://issues.openmrs.org/browse/IDGEN-79)

## SUMMARY:

Link Auto Generation Options EndPoints to OWA

### CODE REVIEW

**Comment by kwahalf** [comment]
@andela-dbamidele  your builds are failing

File: api/src/main/java/org/openmrs/module/idgen/AutoGenerationOption.java [hunk 1]
```diff
@@ -22,7 +22,6 @@
 public class AutoGenerationOption {
 	
 	//***** PROPERTIES *****
-
```
Comment (kwahalf) [inline api/src/main/java/org/openmrs/module/idgen/AutoGenerationOption.java]: why did you change this

File: api/src/main/java/org/openmrs/module/idgen/service/IdentifierSourceService.java [hunk 2]
```diff
@@ -217,7 +217,7 @@
      * @throws non-unique exception if more than one auto-generation option for this type
      */
     @Transactional(readOnly = true)
-    public AutoGenerationOption getAutoGenerationOption(PatientIdentifierType type) throws APIException;
```
Comment (kwahalf) [inline api/src/main/java/org/openmrs/module/idgen/service/IdentifierSourceService.java]: why did you change this

File: omod/TestData.xml [hunk 3]
```diff
@@ -0,0 +1,31 @@
+<?xml version='1.0' encoding='UTF-8'?>
```
Comment (kwahalf) [inline omod/TestData.xml]: why did you add this?

File: omod/src/test/java/org/openmrs/module/idgen/web/RestTestConstants.java [hunk 4]
```diff
@@ -0,0 +1,16 @@
+/**
```
Comment (kwahalf) [inline omod/src/test/java/org/openmrs/module/idgen/web/RestTestConstants.java]: why did you add this?

File: omod/src/test/resources/TestingApplicationContext.xml [hunk 5]
```diff
@@ -0,0 +1,29 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (kwahalf) [inline omod/src/test/resources/TestingApplicationContext.xml]: why did you add this?

**Comment by cnagadya** [comment]
This PR seems to contain some of the IDGEN-71 work. Did you rebase after @kwahalf's PR got merged?

File: omod/src/test/resources/org/openmrs/module/idgen/include/TestData.xml [hunk 6]
```diff
@@ -0,0 +1,31 @@
+<?xml version='1.0' encoding='UTF-8'?>
```
Comment (kwahalf) [inline omod/src/test/resources/org/openmrs/module/idgen/include/TestData.xml]: why did you add this?

File: omod/src/test/resources/test-hibernate.cfg.xml [hunk 7]
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (kwahalf) [inline omod/src/test/resources/test-hibernate.cfg.xml]: why did you add this?

File: owa/app/js/components/manage_autogeneration_option/helpers.js [hunk 8]
```diff
@@ -0,0 +1,80 @@
+import axios from "axios";
```
Comment (kwahalf) [inline owa/app/js/components/manage_autogeneration_option/helpers.js]: why do you need this when you already have apiHelper.js ?

File: owa/webpack.config.js [hunk 9]
```diff
@@ -51,7 +51,7 @@ var getConfig = function () {
 	  } catch (err) {
 	    // create file with defaults if not found
 	    config = {
-	      'LOCAL_OWA_FOLDER': '/Users/denisjuma/Desktop/referenceapplication-standalone-2.6.0/appdata\\owa/',
+	      'LOCAL_OWA_FOLDER': '/Users/andeladeveloper/referenceapplication-standalone-2.6.0/appdata/owa/',
```
Comment (kwahalf) [inline owa/webpack.config.js]: why are you pushing this change

File: owa/webpack.config.js [hunk 10]
```diff
@@ -156,7 +156,7 @@ plugins.push(new CopyWebpackPlugin([{
 var webpackConfig = {
   quiet: false,
   entry: {
-	  app : `${__dirname}/app/js/openmrs-owa-idgen`,
+	  app : ['babel-polyfill', `${__dirname}/app/js/openmrs-owa-idgen`],
```
Comment (kwahalf) [inline owa/webpack.config.js]: any reason for this change?

File: omod/src/test/java/org/openmrs/module/idgen/web/RestTestConstants.java (hunk 4 above)
Comment (andela-dbamidele) [reply omod/src/test/java/org/openmrs/module/idgen/web/RestTestConstants.java]: I think I've added these files before I handed the ticket to you. I'll address them

File: omod/TestData.xml (hunk 3 above)
Comment (andela-dbamidele) [reply omod/TestData.xml]: It came with the rebase.

File: api/src/main/java/org/openmrs/module/idgen/service/IdentifierSourceService.java (hunk 2 above)
Comment (andela-dbamidele) [reply api/src/main/java/org/openmrs/module/idgen/service/IdentifierSourceService.java]: Same for this as well

File: api/src/main/java/org/openmrs/module/idgen/AutoGenerationOption.java (hunk 1 above)
Comment (andela-dbamidele) [reply api/src/main/java/org/openmrs/module/idgen/AutoGenerationOption.java]: I didn't. It was like that when I rebased.

File: owa/app/js/components/manage_autogeneration_option/helpers.js (hunk 8 above)
Comment (andela-dbamidele) [reply owa/app/js/components/manage_autogeneration_option/helpers.js]: I was already using this before @adamsdenniskariuki informed me of`apiHelper.js`. And I find this straightforward and easy to use

File: owa/webpack.config.js (hunk 9 above)
Comment (andela-dbamidele) [reply owa/webpack.config.js]: It wasn't in the gitignore.

File: owa/webpack.config.js (hunk 10 above)
Comment (andela-dbamidele) [reply owa/webpack.config.js]: I added it to enable me use es6 async/await. It provides support for all stage-0 features as well

**Comment by andela-dbamidele** [comment]
I checked the failing tests and it seems they are from the logEntries. I have no idea how this came to be.

**Comment by andela-dbamidele** [comment]
@cnagadya Yes I rebased before I pushed

File: api/src/main/java/org/openmrs/module/idgen/AutoGenerationOption.java (hunk 1 above)
Comment (kwahalf) [reply api/src/main/java/org/openmrs/module/idgen/AutoGenerationOption.java]: It shows as a change from your end

File: api/src/main/java/org/openmrs/module/idgen/service/IdentifierSourceService.java (hunk 2 above)
Comment (kwahalf) [reply api/src/main/java/org/openmrs/module/idgen/service/IdentifierSourceService.java]: it shows as a change from your end

File: omod/TestData.xml (hunk 3 above)
Comment (kwahalf) [reply omod/TestData.xml]: it shows as a change from your end

File: owa/app/js/components/manage_autogeneration_option/helpers.js (hunk 8 above)
Comment (kwahalf) [reply owa/app/js/components/manage_autogeneration_option/helpers.js]: it's a reppitition since it does the same thing

File: omod/src/test/java/org/openmrs/module/idgen/web/RestTestConstants.java (hunk 4 above)
Comment (kwahalf) [reply omod/src/test/java/org/openmrs/module/idgen/web/RestTestConstants.java]: Remove them it's not needed

File: owa/app/js/components/manage_autogeneration_option/helpers.js (hunk 8 above)
Comment (kwahalf) [reply owa/app/js/components/manage_autogeneration_option/helpers.js]: how hard is apiCall(data, requestType, url) compared to this down here?

File: owa/webpack.config.js (hunk 10 above)
Comment (kwahalf) [reply owa/webpack.config.js]: where have we used es6 async/await?

Comment (andela-dbamidele) [reply owa/webpack.config.js]: I used it [here](https://github.com/andela-dbamidele/openmrs-module-idgen/blob/eb84303d2d21bb691550b1b94e1bb8d0648b3294/owa/app/js/components/manage_autogeneration_option/helpers.js)

Comment (kwahalf) [reply owa/webpack.config.js]: The same helpers.js which is a duplicate right?

Comment (andela-dbamidele) [reply owa/webpack.config.js]: Do you mean duplicates to `apiHelper`? They are doing pretty different things.

File: api/src/main/java/org/openmrs/module/idgen/AutoGenerationOption.java (hunk 1 above)
Comment (kwahalf) [reply api/src/main/java/org/openmrs/module/idgen/AutoGenerationOption.java]: If you didn't change them then your commits are mixed with someone else's commits

File: api/src/main/java/org/openmrs/module/idgen/service/IdentifierSourceService.java (hunk 2 above)
Comment (kwahalf) [reply api/src/main/java/org/openmrs/module/idgen/service/IdentifierSourceService.java]: If you didn't change them then your commits are mixed with someone else's commits

File: owa/app/js/components/App.jsx [hunk 11]
```diff
@@ -7,21 +7,22 @@
  * graphic logo is a trademark of OpenMRS Inc.
  */
 import React from 'react';
-import 'bootstrap/dist/css/bootstrap.css'; 
-import Header from './common/header'
-import BodyLayout from './bodyLayout'
-import BreadCrumbs from './breadcrumbs/breadcrumbs'
+import 'bootstrap/dist/css/bootstrap.css';
+import 'sweetalert2/dist/sweetalert2.min.css';
```
Comment (LWanjiru) [inline owa/app/js/components/App.jsx]: Minor: Consider following the react best practices for dependency and local imports. The styling imports should be after the local file imports. This resource explains it better https://engineering.musefind.com/our-best-practices-for-writing-react-components-dec3eb5c3fc8

File: owa/webpack.config.js (hunk 10 above)
Comment (kwahalf) [reply owa/webpack.config.js]: We are going in circles I don't know why. what is the purpose of helpers.js ?

File: owa/app/js/components/App.jsx (hunk 11 above)
Comment (andela-dbamidele) [reply owa/app/js/components/App.jsx]: Yea you are right. I did that because it's an `absolute/package` import and it's a best practice to import them before importing the absolute packages. It's also part of the eslint rules.

File: owa/webpack.config.js (hunk 10 above)
Comment (andela-dbamidele) [reply owa/webpack.config.js]: Instead of making my api calls in my component's componentDidMount, I created the helper to do the api calls for all the values I'll be using in my component. 

I could have done exactly the same thing in my component and it result in a lot of lines.

File: owa/app/js/components/manage_autogeneration_option/helpers.js [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+    .then(
+      res => {
+        return res.data.results;
+      },
+      ({ response }) => {
+        console.log(response);
+      }
+    );
+  const genOptionAutogenerateSources = await axios
+    .get('/openmrs-standalone/ws/rest/v1/idgen/identifiersource')
```
Comment (kwahalf) [inline owa/app/js/components/manage_autogeneration_option/helpers.js]: this will only work on your stand alone. use the apiHelper.js  it handles the url for you

File: owa/app/js/components/manage_autogeneration_option/helpers.js [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+    .then(
+      res => {
+        return res.data.results;
+      },
+      ({ response }) => {
+        console.log(response);
+      }
+    );
+  const identifierTypes = await axios
+    .get('/openmrs-standalone/ws/rest/v1/patientidentifiertype')
```
Comment (kwahalf) [inline owa/app/js/components/manage_autogeneration_option/helpers.js]: same here

File: owa/app/js/components/manage_autogeneration_option/helpers.js [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+    .then(
+      res => {
+        return res.data.results;
+      },
+      ({ response }) => {
+        console.log(response);
+      }
+    );
+  const genOptions = await axios
+    .get('/openmrs-standalone/ws/rest/v1/idgen/autogenerationoption')
```
Comment (kwahalf) [inline owa/app/js/components/manage_autogeneration_option/helpers.js]: same here

File: owa/app/js/components/manage_autogeneration_option/helpers.js [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+    .then(
+      async res => {
+        const results = res.data.results;
+        if (results.length > 0) {
+          const allGenOptions = [];
+
+          results.map(result => {
+            const promise = axios.get(
+              `/openmrs-standalone/ws/rest/v1/idgen/autogenerationoption/${result.uuid}`
```
Comment (kwahalf) [inline owa/app/js/components/manage_autogeneration_option/helpers.js]: same here, just discard this whole file and use the apiHelper.js

**Comment by andela-dbamidele** [comment]
@kwahalf The changes you requested have been made.

File: owa/webpack.config.js [hunk 16]
```diff
@@ -179,7 +179,7 @@ var webpackConfig = {
 	    loader: 'babel-loader',
 	    exclude: /node_modules/,
 	    query: {
-	        presets: [ 'es2015', 'react' ],
+	        presets: [ 'es2015', 'react', 'stage-0' ],
```
Comment (kwahalf) [inline owa/webpack.config.js]: do we still need this change?

File: owa/webpack.config.js (hunk 10 above)
Comment (kwahalf) [reply owa/webpack.config.js]: do we still need this?

File: api/src/main/java/org/openmrs/module/idgen/AutoGenerationOption.java (hunk 1 above)
Comment (kwahalf) [reply api/src/main/java/org/openmrs/module/idgen/AutoGenerationOption.java]: remove this file from your commit

Comment (dkayiwa) [reply api/src/main/java/org/openmrs/module/idgen/AutoGenerationOption.java]: Your commit should not have any changes to this file.

File: api/src/main/java/org/openmrs/module/idgen/service/IdentifierSourceService.java (hunk 2 above)
Comment (dkayiwa) [reply api/src/main/java/org/openmrs/module/idgen/service/IdentifierSourceService.java]: Your commit should not have any changes to this file.

**Comment by andela-dbamidele** [comment]
All changes made.

File: owa/app/js/components/manage_autogeneration_option/ManageAutoGenerationOption.jsx [hunk 17]
```diff
@@ -1,52 +1,33 @@
...
+  Col,
+  FormGroup,
+  Label,
+  Input,
+  Table,
+  Modal,
+  ModalHeader,
+  ModalBody,
+  ModalFooter,
+  Alert
```
Comment (kwahalf) [inline owa/app/js/components/manage_autogeneration_option/ManageAutoGenerationOption.jsx]: is this formatting standard?

File: owa/app/js/components/manage_autogeneration_option/ManageAutoGenerationOption.jsx [hunk 18]
```diff
@@ -79,47 +107,118 @@ class ManageAutoGenerationOption extends Component {
...
-      modal: false
+    const {
+      id,
+      uuid,
+      identifierType,
+      currentLocation,
+      autogenerateSource,
+      autoGenerationEnabled,
+      manualEntryEnabled
+    } = currentGenOption;
```
Comment (kwahalf) [inline owa/app/js/components/manage_autogeneration_option/ManageAutoGenerationOption.jsx]: same here

File: owa/app/js/components/manage_autogeneration_option/ManageAutoGenerationOption.jsx [hunk 19]
```diff
@@ -173,25 +271,43 @@ class ManageAutoGenerationOption extends Component {
...
-        genOptions
-      }, () => {
+      const {
+        id,
+        identifierType,
+        currentLocation,
+        autogenerateSource,
+        autoGenerationEnabled,
+        manualEntryEnabled
+      } = currentGenOption;
```
Comment (kwahalf) [inline owa/app/js/components/manage_autogeneration_option/ManageAutoGenerationOption.jsx]: same here

File: owa/app/js/components/manage_autogeneration_option/ManageAutoGenerationOption.jsx (hunk 17 above)
Comment (andela-dbamidele) [reply owa/app/js/components/manage_autogeneration_option/ManageAutoGenerationOption.jsx]: Yes. It's what I've been doing and I think its eslint acceptable way of doing it. But I'm open to any other suggestion or better way of doing it.

File: owa/app/js/components/manage_autogeneration_option/ManageAutoGenerationOption.jsx (hunk 18 above)
Comment (andela-dbamidele) [reply owa/app/js/components/manage_autogeneration_option/ManageAutoGenerationOption.jsx]: Same as above

File: owa/app/js/components/manage_autogeneration_option/ManageAutoGenerationOption.jsx (hunk 19 above)
Comment (andela-dbamidele) [reply owa/app/js/components/manage_autogeneration_option/ManageAutoGenerationOption.jsx]: Same as above

**Comment by dkayiwa** [comment]
@andela-dbamidele did you remove the changes that @cnagadya talked about?

**Comment by andela-dbamidele** [comment]
Yes I did. @dkayiwa
